src/operator/tensor/indexing_op-inl.cuh

```diff
@@ -0,0 +1,289 @@
+/*!
+ * Copyright (c) 2017 by Contributors
+ * \file indexing_op-inl.cuh
+ * \brief CUDA implementations for indexing_op.h
+ * \author Antti-Pekka Hynninen
+*/
+#ifndef MXNET_OPERATOR_TENSOR_INDEXING_OP_CUH_
+#define MXNET_OPERATOR_TENSOR_INDEXING_OP_CUH_
+#include <cub/device/device_run_length_encode.cuh>
+#include <cub/device/device_scan.cuh>
+
+namespace mxnet {
+namespace op {
+const int kWarpSize = 32;
+
+template<int SZ, typename DType, typename IdxType>
+__global__ void AddTakeGradLargeBatchKernel(DType* dst,
+                                           // If idx_start == NULL, then in-kernel edge
+                                           // detection is used
+                                           const IdxType *idx_start,
+                                           // idx_start_size_ptr ignored if idx_start == NULL
+                                           const int* idx_start_size_ptr,
+                                           const IdxType *sorted, const IdxType *index,
+                                           const DType *src,
+                                           int ymax, int xmax) {
+  // Size of the shared memory is [blockDim.x*SZ*blockDim.y]*sizeof(DType)
+  extern __shared__ char sh_grad_weight_char[];
+  DType* sh_grad_weight = (DType*)sh_grad_weight_char;
+
+  int iidx_end = (idx_start == NULL) ? ymax : *idx_start_size_ptr;
+
+  for (int iidx = blockIdx.y;iidx < iidx_end;iidx += gridDim.y) {
+
+    // Thread block sums up elements in the range [idx_begin, idx_end-1]
+    int idx_begin, idx_end;
+    int sorted_value;
+    if (idx_start == NULL) {
+      idx_begin = iidx;
+      sorted_value = static_cast<int>(sorted[idx_begin]);

```

Comment:
I find that if the IdxType is float, directly casting may be problematic for values like 9.99999. Should we try to solve it? The previous code has also not handled this type of special case.